lib/spack/spack/environment.py

```diff
@@ -261,132 +261,152 @@ def apply_modifications(self):
+        env_after = json.loads(output)
 
         # If we're in python2, convert to str objects instead of unicode
         # like json gives us.  We can't put unicode in os.environ anyway.
         if sys.version_info[0] < 3:
-            after_source_env = dict((k.encode('utf-8'), v.encode('utf-8'))
-                                    for k, v in after_source_env.items())
+            env_after = dict((k.encode('utf-8'), v.encode('utf-8'))
+                             for k, v in env_after.items())
 
         # Filter variables that are not related to sourcing a file
-        to_be_filtered = 'SHLVL', '_', 'PWD', 'OLDPWD'
-        for d in after_source_env, this_environment:
+        to_be_filtered = 'SHLVL', '_', 'PWD', 'OLDPWD', 'PS2'
+        for d in env_after, env_before:
             for name in to_be_filtered:
                 d.pop(name, None)
 
         # Fill the EnvironmentModifications instance
+        env = EnvironmentModifications()
 
         # New variables
-        new_variables = set(after_source_env) - set(this_environment)
-        for x in new_variables:
-            env.set(x, after_source_env[x])
+        new_variables = set(env_after) - set(env_before)
         # Variables that have been unset
-        unset_variables = set(this_environment) - set(after_source_env)
-        for x in unset_variables:
-            env.unset(x)
+        unset_variables = set(env_before) - set(env_after)
         # Variables that have been modified
         common_variables = set(
-            this_environment).intersection(set(after_source_env))
+            env_before).intersection(set(env_after))
         modified_variables = [x for x in common_variables
-                              if this_environment[x] != after_source_env[x]]
+                              if env_before[x] != env_after[x]]
 
-        def return_separator_if_any(first_value, second_value):
+        def return_separator_if_any(*args):
             separators = ':', ';'
             for separator in separators:
-                if separator in first_value and separator in second_value:
-                    return separator
-            return None
+                for arg in args:
+                    if separator in arg:
+                        return separator
 
-        for x in modified_variables:
-            current = this_environment[x]
-            modified = after_source_env[x]
-            sep = return_separator_if_any(current, modified)
-            if sep is None:
-                # We just need to set the variable to the new value
-                env.set(x, after_source_env[x])
+        # Add variables to env.
+        # Assume that variables with 'PATH' in the name or that contain
+        # separators like ':' or ';' are more likely to be paths
+        for x in new_variables:
+            sep = return_separator_if_any(env_after[x])
+            if sep or 'PATH' in x:
+                env.prepend_path(x, env_after[x])
```

Comment:
I agree. Does `prepend_path` support this right now? If not, I don't want to make too many more changes in this PR.